sips/sip-021/sip-021-trustless-two-way-peg-for-bitcoin.md

```diff
@@ -0,0 +1,1385 @@
+height is unchanged.
+
+* When mining off of unblessed block-commits, the highest
+frozen block-commit at Bitcoin height N will either be the same as the one in
+Bitcoin height N+1, or it will be the parent of the one in Bitcoin height N+1.
+This is _not true_ when considering blessed block-commits.
+
+### Rules
+
+By introducing both lifecycle phases and Stacker blessings, this proposal
+creates a new fork-choice rule for Stacks: the canonical fork is always tipped
+by the highest descendant of the highest frozen block-commit's Stacks block,
+even if the block for it is not yet published.
+
+In doing so, the system creates a new miner rule: a miner can mine off of any
+descendant of the highest frozen block-commit (i.e. any unblessed
+block-commit), and it can mine off of any descendant of a blessed block-commit.
+Miners freely and individually choose which member of this set of block-commits
+to mine off of.  A miner cannot mine off of any other block-commits.
+
+### Finalization
+
+Once a block-commit receives 100 Bitcoin confirmations, it can no longer be
+blessed.
+
+When Stackers fulfill a peg-out request, they invalidate all Stacks forks in
+which the peg-out would result in an inconsistent chain state (i.e. someone's
+sBTC balance becomes negative).  These forks will never be treated as canonical
+as long as they contain such inconsistencies.  As such, the act of Stackers
+fulfilling a peg-out request that materializes in Stacks block B is declaring
+that any Stacks fork that descends from B might be canonical (i.e. the
+canonical fork must include B).
+
+Consequently, the act of processing a peg-out is used to keep the chain alive
+even if someone manages to mine 99 block-commits without disclosing their
+blocks.  As long as there exists at least one valid publicly-disclosed block,
+then Stackers can use it to authenticate and process peg-outs and in doing so,
+require that block to be part of the canonical chain.  This enables miners and
+Stackers to prevent a hidden-block miner from halting the chain while spending
+far fewer resources than the attacker.  It also prevents a 51% attacker from
+denying sBTC transfers and peg-outs: if any sBTC transfer materializes in the
+honest fork, and Stackers process at least one peg-out on that fork, then the
+51% attacker's fork is obliged to _also_ process the _same_ sBTC transfers in
+order to keep their fork canonical.
+
+# Related Work
+
+This section is not done yet.  Notes follow.
+
+* PowPeg in RSK
+
+	* Requires HSMs on a majority of Bitcoin miners (brittle;
+closed-membership)
+
+	* Has a token with coinbases to incentivize validator
+liveness regardless of tx volume (good!)
+
+	* Probably the closest system to what
+we're building
+
+* Sidechains / drivechains
+
+	* No incentive for validators to
+process transactions, outside of miniscule tx fees
+
+	* No compelling answer for
+dealing with sidechain reorgs (Bitcoin miners have to fully validate, which
+does not scale)
```

Comment:
> * No compelling answer for dealing with sidechain reorgs (Bitcoin miners have to fully validate, which does not scale)

It's worth separating out these two claims. Although adequately dealing with re-orgs does require the bitcoin miners (who are also the custodians of BTC locked in the hashrate escrow) to fully validate the chain so they know which pegout requests to process, the issue of scaling full validation would still exist even if the question of how to handle re-orgs was adequately addressed. So I will address each claim separately:

> Bitcoin miners have to fully validate, which does not scale

Using "proof-sync" clients such as [Zerosync](https://zerosync.org/), the marginal effort of validating an extra chain would be minimal, enabling the trustless scaling of full validation.

> No compelling answer for dealing with sidechain reorgs

The general solution here is to wait some number of confirmations before processing peg-outs. If there is a re-org after said number of confirmations then yes, the peg-outs could be double-spent, causing the sidechain to become undercollateralized.

One simple way to deal with this possibility is to require that re-orgs are not allowed at a block depth equal to or greater than the block depth at which peg-outs will be processed. IIUC Stacks/sBTC does something like this:

> If Stackers have received their peg-wallet hand-off for this reward cycle and fail to fulfill a request within 50 Bitcoin blocks of the request being finalized (i.e. at most 150 Bitcoin blocks after the request is submitted), then the system transitions to Recovery mode and PoX payouts are repurposed for fulfilling pending peg-out requests.

I don't know if these answers meet the author's bar for "compelling" but I find them reasonable given the inherent tradeoffs of the sidechain approach.

Reply:
> Using "proof-sync" clients such as Zerosync, the marginal effort of validating an extra chain would be minimal, enabling the trustless scaling of full validation.

A few questions:
* Is Zerosync live yet?  If not, then our point stands at least until it's live.
* If Zerosync is live, does it apply to sidechains with computation models that are at least as expressive as Clarity?  If not, then Bitcoin's ability to scale through STARK proofs for sidechains does not help users achieve expressive smart contracts for Bitcoin (meaning, zerosync's existence does not invalidate the novelty of our work).
* Is Zerosync proof generation open-membership?  As in, can multiple prover coalitions exist?  If so, can two coalitions of provers submit conflicting proofs on the state of the sidechain?  If so, then can Bitcoin determine which proof represents the majority coalition?  If the answer to any of these three criteria is "no," then it's also strictly worse than our work.  A sidechain whose state is decided by a fixed set of provers has all the problems that PoS does (for example, it's impossible to dislodge or overpower malicious provers in-band, and a 33% or larger malicious coalition can _halt_ the sidechain).

> One simple way to deal with this possibility is to require that re-orgs are not allowed at a block depth equal to or greater than the block depth at which peg-outs will be processed. IIUC Stacks/sBTC does something like this:

If the sBTC wallet is ever undercollateralized, then Recovery Mode engages and PoX payouts are repurposed to restore the missing BTC.  No sidechain does this that I at least am aware of.

Reply:
> - Is Zerosync live yet? If not, then our point stands at least until it's live.
> - If Zerosync is live, does it apply to sidechains with computation models that are at least as expressive as Clarity? If not, then Bitcoin's ability to scale through STARK proofs for sidechains does not help users achieve expressive smart contracts for Bitcoin (meaning, zerosync's existence does not invalidate the novelty of our work).

There are no production-grade proof-sync clients for bitcoin-like chains yet. However, there are proof-sync clients for fully-expressive EVM and Cairo chains. These are implemented as smart contracts and used for verifying the state of validity rollups built with those computation models. Some example include:
- EVM: Polygon zkEVM, zkSync Era
- Cairo: Starknet

> Is Zerosync proof generation open-membership? As in, can multiple prover coalitions exist? 

Yes

> If so, can two coalitions of provers submit conflicting proofs on the state of the sidechain?

Yes, but... 

> If so, then can Bitcoin determine which proof represents the majority coalition?

Short answer: bitcoin consensus does not verify the proof. Only the miners need to verify the proof, so they know which drivechain blocks to build on and which withdrawals (what you called "redemptions" in another comment) to upvote (or downvote). It doesn't matter how many proofs get thrown at them by how many diffent "coalitions" of provers, the miners can use the proofs to figure out what they need to know so that they only allow valid withdrawals to go through.